synapse/federation/federation_server.py

```diff
@@ -135,22 +141,41 @@ async def on_incoming_transaction(
         request_time = self._clock.time_msec()
 
         transaction = Transaction(**transaction_data)
+        transaction_id = transaction.transaction_id  # type: ignore
 
-        if not transaction.transaction_id:  # type: ignore
+        if not transaction_id:
             raise Exception("Transaction missing transaction_id")
 
-        logger.debug("[%s] Got transaction", transaction.transaction_id)  # type: ignore
+        logger.debug("[%s] Got transaction", transaction_id)
 
-        # use a linearizer to ensure that we don't process the same transaction
-        # multiple times in parallel.
+        # We wrap in a ResponseCache so that we de-duplicate retried
+        # transactions.
+        return await self._transaciton_resp_cache.wrap(
+            (origin, transaction_id),
+            self._on_incoming_transaction_inner,
+            origin,
+            transaction,
+            request_time,
+        )
+
+    async def _on_incoming_transaction_inner(
+        self, origin: str, transaction: Transaction, request_time: int
+    ) -> Tuple[int, Dict[str, Any]]:
+        # Use a linearizer to ensure that process transactions from a remote in
+        # order.
         with (
             await self._transaction_linearizer.queue(
                 (origin, transaction.transaction_id)  # type: ignore
             )
         ):
-            result = await self._handle_incoming_transaction(
-                origin, transaction, request_time
-            )
+            # We rate limit here *after* we've queued up the incoming requests,
+            # so that we don't fill up the ratelimiter with blocked requests.
```

Comment:
I'm not sure I'm following what this change is. Could you describe it a bit more? I see how the rate-limiter is moved deeper into the stack, but I'm not sure how a "blocked request" would fill it up previously that won't now? Does blocked mean a 429 is returned or that it got stuck in the linearizer?